consensus/view_change.go

```diff
@@ -214,228 +191,72 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
+	if err != nil {
+		consensus.getLogger().Error().Err(err).
+			Uint64("viewID", recvMsg.ViewID).
+			Uint64("blockNum", recvMsg.BlockNum).
+			Str("msgSender", recvMsg.SenderPubkey.Bytes.Hex()).
+			Msg("Verify View Change Message Error")
+		return
 	}
 
-	// m2 type message
-	if !hasBlock {
-		_, ok := consensus.nilSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]
-		if ok {
-			consensus.getLogger().Debug().
-				Str("validatorPubKey", senderKey.Bytes.Hex()).
-				Msg("[onViewChange] Already Received M2 message from validator")
-			return
-		}
-
-		if !recvMsg.ViewchangeSig.VerifyHash(senderKey.Object, NIL) {
-			consensus.getLogger().Warn().Msg("[onViewChange] Failed To Verify Signature For M2 Type Viewchange Message")
-			return
-		}
-
-		consensus.getLogger().Info().
-			Str("validatorPubKey", senderKey.Bytes.Hex()).
-			Msg("[onViewChange] Add M2 (NIL) type message")
-		consensus.nilSigs[recvMsg.ViewID][senderKey.Bytes.Hex()] = recvMsg.ViewchangeSig
-		consensus.nilBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true) // Set the bitmap indicating that this validator signed.
-	} else { // m1 type message
-		if consensus.BlockVerifier(preparedBlock); err != nil {
-			consensus.getLogger().Error().Err(err).Msg("[onViewChange] Prepared block verification failed")
-			return
-		}
-		_, ok := consensus.bhpSigs[recvMsg.ViewID][senderKey.Bytes.Hex()]
-		if ok {
-			consensus.getLogger().Debug().
-				Str("validatorPubKey", senderKey.Bytes.Hex()).
-				Msg("[onViewChange] Already Received M1 Message From the Validator")
-			return
-		}
-		if !recvMsg.ViewchangeSig.VerifyHash(recvMsg.SenderPubkey.Object, recvMsg.Payload) {
-			consensus.getLogger().Warn().Msg("[onViewChange] Failed to Verify Signature for M1 Type Viewchange Message")
-			return
-		}
-
-		// first time receive m1 type message, need verify validity of prepared message
-		if len(consensus.m1Payload) == 0 || !bytes.Equal(consensus.m1Payload, recvMsg.Payload) {
-			if len(recvMsg.Payload) <= 32 {
-				consensus.getLogger().Warn().
-					Int("len", len(recvMsg.Payload)).
-					Msg("[onViewChange] M1 RecvMsg Payload Not Enough Length")
-				return
-			}
-			blockHash := recvMsg.Payload[:32]
-			aggSig, mask, err := consensus.ReadSignatureBitmapPayload(recvMsg.Payload, 32)
-			if err != nil {
-				consensus.getLogger().Error().Err(err).Msg("[onViewChange] M1 RecvMsg Payload Read Error")
-				return
-			}
-
-			if !consensus.Decider.IsQuorumAchievedByMask(mask) {
-				consensus.getLogger().Warn().
-					Msgf("[onViewChange] Quorum Not achieved")
-				return
-			}
-
-			// Verify the multi-sig for prepare phase
-			if !aggSig.VerifyHash(mask.AggregatePublic, blockHash[:]) {
-				consensus.getLogger().Warn().
-					Hex("blockHash", blockHash).
-					Msg("[onViewChange] failed to verify multi signature for m1 prepared payload")
-				return
+	if msgType == M1 {
+		preparedMsgs := consensus.FBFTLog.GetMessagesByTypeSeq(
```

Comment:
For the viewChange's M1 msg. The block is parsed from the message. Not from local log, the FBFTLog is for self commit of viewChange (should be in InitPayload). Can you double check?

Reply:
will double check.

Reply:
The original logic also add a new prepared message to FBFTLog if the validator can't find it locally.  So this validator has prepared message, but it received view change with M1.

![image](https://user-images.githubusercontent.com/4513927/94977329-1d38ca80-04cd-11eb-88ea-560b5facaa1f.png)

Reply:
The original logic only add the m1 payload as fbft log if the payload is never seen before. Did you keep that logic?

Reply:
discussed in the new PR.